pinc/page_table.inc

```diff
@@ -8,6 +8,215 @@ include_once($relPath.'../tools/proofers/PPage.inc'); // url_for_pi_do_particula
+{
+    $project_round = get_Round_for_project_state($project->state);
+
+    // This project may have skipped some rounds, and/or it may have rounds
+    // yet to do, so there may be some round-columns with no data in them.
+    // Figure out which ones to display.
+    //
+    $rounds_to_display = get_rounds_with_data($project->projectid);
+
+    // If the project is in a round, then users expect that round to appear here,
+    // even if there's no data in it.
+    //
+    if ( !is_null($project_round) )
+    {
+        // Look for $project_round in $rounds_to_display.
+        // If it does appear, it will normally be the last item,
+        // but check everywhere, just in case.
+        $found = FALSE;
+        foreach ( $rounds_to_display as $round )
+        {
+            if ( $round->id == $project_round->id )
+            {
+                $found = TRUE;
+                break;
+            }
+        }
+        if ( !$found )
+        {
+            $rounds_to_display[] = $project_round;
+        }
+    }
+
+    return $rounds_to_display;
+}
+
+/*
+Generator to return page table data for a project, one page at a time.
+Data is returned in an multi-level associative array that abstracts away
+the lower-level database structure. The returned format looks like:
+
+{
+    "image": "018.png",
+    "image_url": "https://www.pgdp.net/projects/projectID460978cb8e116/018.png",
+    "state": "P3.page_saved",
+    "image_size": 45541,
+    "pagerounds": {
+        "OCR": {
+            "page_size": "1445"
+        },
+        "P1": {
+            "page_size": "1446",
+            "is_diff": "0",
+            "username": "username",
+            "user_page_tally": "117"
+            "wordcheck_ran": "0",
+            "modified_timestamp": "1175026200",
+        },
+        "P2": {
+            ...
+        },
+        ...
+    }
+}
+*/
+function fetch_page_table_data($project, $page_selector = NULL, $proofed_in_round = NULL)
+{
+    $rounds_to_display = get_rounds_to_display($project);
+
+    $fields_to_get = "{$project->projectid}.image AS image, length(master_text), state";
+    $tables = $project->projectid;
+    $prev_text_column_name = 'master_text';
+    foreach ( $rounds_to_display as $round )
+    {
+        $rn = $round->round_number;
+
+        $tallyboard = new TallyBoard( $round->id, 'U' );
+        list($joined_with_user_page_tallies,$user_page_tally_column) =
+            $tallyboard->get_sql_joinery_for_current_tallies( "users$rn.u_id" );
+
+        if(is_formatting_round($round))
+        {
+            $wordcheck_query = "NULL as wordcheck_status$rn";
+        }
+        else
+        {
+            $wordcheck_query = "
+                (
+                    SELECT count(*)
+                    FROM wordcheck_events
+                    WHERE projectid = '$project->projectid' AND
+                        image = {$project->projectid}.image AND
+                        username = $round->user_column_name AND
```

Comment:
do we want to use the sprintf and encoding here

Reply:
Yes & done.